src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,476 @@
+            {
+                return ParseNumber(token);
+            }
+
+            return new JsonTextNode(
+                token.With(kind: JsonKind.TextToken).AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    string.Format(WorkspacesResources._0_unexpected, firstChar.Char),
+                    firstChar.Span)));
+        }
+
+        private JsonConstructorNode ParseConstructor(JsonToken token)
+        {
+            var newKeyword = token.With(kind: JsonKind.NewKeyword);
+            var nameToken = ConsumeToken(JsonKind.TextToken, WorkspacesResources.Name_expected);
+            var openParen = ConsumeToken(JsonKind.OpenParenToken, _openParenExpected);
+
+            var savedInConstructor = _inConstructor;
+            _inConstructor = true;
+
+            var result = new JsonConstructorNode(
+                newKeyword,
+                nameToken,
+                openParen,
+                ParseSequence(),
+                ConsumeToken(JsonKind.CloseParenToken, _closeParenExpected));
+
+            _inConstructor = savedInConstructor;
+            return result;
+        }
+
+        private bool TryMatch(JsonToken token, string val, JsonKind kind, out JsonKind newKind)
+        {
+            if (Matches(token, val))
+            {
+                newKind = kind;
+                return true;
+            }
+
+            newKind = default;
+            return false;
+        }
+
+        private bool Matches(JsonToken token, string val)
+        {
+            var chars = token.VirtualChars;
+            if (chars.Length != val.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0, n = val.Length; i < n; i++)
+            {
+                if (chars[i].Char != val[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool IsDigit(char ch)
+            => ch >= '0' && ch <= '9';
+
+        private JsonLiteralNode ParseLiteral(JsonToken textToken, JsonKind kind)
+            => new JsonLiteralNode(textToken.With(kind: kind));
+
+        private JsonValueNode ParseNumber(JsonToken textToken)
+        {
+            var numberToken = textToken.With(kind: JsonKind.NumberToken);
+            return new JsonLiteralNode(numberToken);
+        }
+
+        private JsonEmptyValueNode ParseEmptyValue()
+            => new JsonEmptyValueNode(ConsumeCurrentToken());
+
+        private JsonArrayNode ParseArray()
+        {
+            var savedInArray = _inArray;
+            _inArray = true;
+
+            var result = new JsonArrayNode(
+                ConsumeCurrentToken(),
+                ParseSequence(),
+                ConsumeToken(JsonKind.CloseBracketToken, _closeBracketExpected));
+
+            _inArray = savedInArray;
+            return result;
+        }
+
+        private JsonObjectNode ParseObject()
+        {
+            var savedInObject = _inObject;
```

Comment:
This seems wrong, and I'm surprised this would not cause bugs.
I'd expect that we're in a single context at a given time (either object, array or constructor). But it looks like this invariant is not protected.
I would have thought the other two properties would have been saved and reset to `false`. 
Alternatively, I think we should use an enum for context (None, Object, Array, Constructor).

Reply:
> I'd expect that we're in a single context at a given time (either object, array or constructor). 

Nope.  This works like how things work in the C#/TS parsers.  You have a stack of contexts that you're in, and you're pushing/popping from that stack.  However, for efficiency, that stack is stored as an int in C#/TS, and i just used 3 bools here.

THe reason you don't actually need an unbounded stack is that you only need to keep track if you're in some type off construct the first time you enter it.  If you recursively enter that same construct again, you don't actually need to record that.  

Then, this information is simply used to know what you should do if you encounter something unexpected during parsing.  For example, if you see.    ```{ ... [ ... }```, i.e. we see a close brace, while in the middle of parsing an array.  Because we know we're *also* in an object, we know we need to bail out of parsing the array so that the object we're parsing can consume hte close brace.